src/MonitorProfiler/ProbeInstrumentation/AssemblyProbePrep.cpp

```diff
@@ -0,0 +1,480 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "corhlpr.h"
+#include "AssemblyProbePrep.h"
+#include "../Utilities/TypeNameUtilities.h"
+#include "../Utilities/MemoryUtilities.h"
+#include "../Utilities/MetadataEnumCloser.h"
+#include "../Utilities/StringUtilities.h"
+
+using namespace std;
+
+#define ENUM_BUFFER_SIZE 10
+#define STRING_BUFFER_LEN 256
+
+AssemblyProbePrep::AssemblyProbePrep(ICorProfilerInfo12* profilerInfo, FunctionID probeFunctionId) :
+    m_pCorProfilerInfo(profilerInfo),
+    m_resolvedCorLibId(0),
+    m_probeFunctionId(probeFunctionId),
+    m_didHydrateProbeCache(false)
+{
+}
+
+bool AssemblyProbePrep::TryGetAssemblyPrepData(ModuleID moduleId, shared_ptr<AssemblyProbePrepData>& data)
+{
+    auto const& it = m_assemblyProbeCache.find(moduleId);
+    if (it != m_assemblyProbeCache.end())
+    {
+        data = it->second;
+        return true;
+    }
+
+    return false;
+}
+
+HRESULT AssemblyProbePrep::PrepareAssemblyForProbes(ModuleID moduleId)
+{
+    HRESULT hr;
+
+    auto const& it = m_assemblyProbeCache.find(moduleId);
+    if (it != m_assemblyProbeCache.end())
+    {
+        return S_OK;
+    }
+
+    COR_LIB_TYPE_TOKENS corLibTypeTokens = {};
+    IfFailRet(EmitNecessaryCorLibTypeTokens(moduleId, corLibTypeTokens));
+
+    mdMemberRef probeMemberRef;
+    IfFailRet(EmitProbeReference(moduleId, probeMemberRef));
+
+    shared_ptr<AssemblyProbePrepData> data(new (nothrow) AssemblyProbePrepData(probeMemberRef, corLibTypeTokens));
```

Comment:
We handle OOM scenarios here but we will throw bad_alloc on stl unordered map operations if there isn't enough memory for allocations. One option here:
- Use throwing new for single allocations
- Use no throw for array like operations, since these could fail if you are requesting larger blocks